### PR TITLE
[AArch64][GlobalISel] Improve non-SVE popcount for 32bit and 64 bit using udot

### PR DESCRIPTION
**Pull request by tgymnich**
Follow up for #95881

Use udot instead of a sequence of uaddlp instructions when summing up lanes for popcount.

### CODE REVIEW

**Review by davemgreen — COMMENTED** [review]
Sounds good, thanks. Can we add global-isel run lines to the existing tests?

**Review by davemgreen — COMMENTED** [review]
It sounds like the tests will need to be rebased once the other patch is committed, but this looks OK to me.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include "AArch64LegalizerInfo.h"
 #include "AArch64RegisterBankInfo.h"
 #include "AArch64Subtarget.h"
+#include "MCTargetDesc/AArch64MCTargetDesc.h"
```
[the remark this answers is not in the document]
Comment (tgymnich) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:17]: My vscode inserted it when using `AArch64::G_UDOT`. But "AArch64Subtarget.h` already transitively includes `MCTargetDesc/AArch64MCTargetDesc.h`.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 2]
```diff
@@ -1904,6 +1905,31 @@ bool AArch64LegalizerInfo::legalizeCTPOP(MachineInstr &MI,
   auto CTPOP = MIRBuilder.buildCTPOP(VTy, Val);
 
   // Sum across lanes.
+
+  if (ST->hasDotProd() && Ty.isVector() && Ty.getNumElements() >= 2 &&
+      Ty.getScalarSizeInBits() != 16) {
+    LLT Dt = Ty == LLT::fixed_vector(2, 64) ? LLT::fixed_vector(4, 32) : Ty;
+    auto Zeros = MIRBuilder.buildConstant(Dt, 0);
+    auto Ones = MIRBuilder.buildConstant(VTy, 1);
+    MachineInstrBuilder SUM;
```
[the remark this answers is not in the document]
Comment (tgymnich) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp]: fixed

File: llvm/test/CodeGen/AArch64/popcount.ll [hunk 3]
```diff
@@ -67,25 +140,338 @@ Entry:
...
+; DOT-GISEL-NEXT:    mov v0.16b, v1.16b
+; DOT-GISEL-NEXT:    ret
+;
+; SVE-GISEL-LABEL: popcount4x32:
+; SVE-GISEL:       // %bb.0: // %Entry
+; SVE-GISEL-NEXT:    cnt v0.16b, v0.16b
+; SVE-GISEL-NEXT:    uaddlp v0.8h, v0.16b
+; SVE-GISEL-NEXT:    uaddlp v0.4s, v0.8h
+; SVE-GISEL-NEXT:    ret
+; NEON-LABEL: popcount4x32:
```
[the remark this answers is not in the document]
Comment (tgymnich) [reply llvm/test/CodeGen/AArch64/popcount.ll]: done

**Comment by tgymnich** [comment]
@davemgreen @aemerson are we ready to merge?

**Comment by davemgreen** [comment]
If you can merge in #95881 and rebase this on top then I think it looks good to me.

**Comment by tgymnich** [comment]
@davemgreen I rebased on main and merged the tests.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**